[Site Isolation] Begin to fix the back forward list for same-origin iframe navigations

charliewolfe opened the pull request:
#### 1085dea9ba4113642ffc393e3f9ac06a778ff26c
<pre>
[Site Isolation] Begin to fix the back forward list for same-origin iframe navigations
<a href="https://bugs.webkit.org/show_bug.cgi?id=274419">https://bugs.webkit.org/show_bug.cgi?id=274419</a>
<a href="https://rdar.apple.com/128420789">rdar://128420789</a>

Reviewed by Alex Christensen.

Some back forward list state for child frames is currently stored only in the web process. This needs to
change with site isolation because web processes can only have history items corresponding to frames they
are hosting or have embedded. With this patch, whenever a root child frame is embedded, we add a separate
item to the back forward list in the UI process. This way, when navigating back or forward, iframe
processes can request history items containing state about same-origin frames that have been embedded by
another process without receiving information about the entire frame tree. We also need to keep track of
which back forward items correspond to initial root child frame loads. This is necessary to determine
which item to navigate an isolated iframe to, while also ignoring these items when the main frame is
moved back or forward.

Further work is required to make the back forward list behave correctly after cross-origin iframe
navigations. More details below.

* Source/WebCore/history/BackForwardClient.h:
* Source/WebCore/history/BackForwardController.cpp:
(WebCore::BackForwardController::addItem):
* Source/WebCore/history/BackForwardController.h:
* Source/WebCore/loader/EmptyClients.cpp:
* Source/WebCore/loader/HistoryController.cpp:
(WebCore::HistoryController::goToItem):
(WebCore::HistoryController::updateBackForwardListClippedAtTarget):
(WebCore::HistoryController::pushState):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::goToItem):
* Source/WebCore/page/Page.h:

Update `addItem` to include a FrameIdentifier parameter so the UI process knows which frame is being
targeted. Also remove several unneeded local frame downcasts.

* Source/WebKit/Shared/WebBackForwardListItem.h:
(WebKit::WebBackForwardListItem::setIsRootChildFrameItem):
(WebKit::WebBackForwardListItem::isRootChildFrameItem const):

Add `m_isRootChildFrameItem` to indicate when a back forward item was created by an initial root child
frame load.

* Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
(WebKit::ProvisionalPageProxy::backForwardAddItem):
* Source/WebKit/UIProcess/ProvisionalPageProxy.h:

* Source/WebKit/UIProcess/WebBackForwardList.cpp:
(WebKit::itemSkippingBackForwardItemsAddedByJSWithoutUserGesture):

Skip root child frame items to avoid targeting initial child frame loads when trying to move the main
frame back/forward. This logic probably still isn’t correct when going back/forward between
cross-origin iframes.

* Source/WebKit/UIProcess/WebFrameProxy.h:
(WebKit::WebFrameProxy::setHasPendingBackForwardItem):
(WebKit::WebFrameProxy::takeHasPendingBackforwardItem):

Add `m_hasPendingBackForwardItem` to indicate when a root frame is waiting for its initial back forward
item to be created.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::goToBackForwardItem):

Change this function to send back forward list items corresponding to isolated iframes to the process
that navigated. `m_lastProcessIdentifier` on `WebBackForwardListItem` is already used for process
selection for PSON. It will initially be the identifier of the web process that constructed the object,
so we can use it for iframe isolation too.

(WebKit::WebPageProxy::didChangeBackForwardList):

Do not call `didChangeBackForwardList` delegates or `setCanGoBack/Forward` for root child frame items to
keep the same behavior as without site isolation.

(WebKit::WebPageProxy::continueNavigationInNewProcess):

Set `m_hasPendingBackForwardItem` to true when creating a new root child frame.

(WebKit::WebPageProxy::backForwardAddItem):
(WebKit::WebPageProxy::backForwardAddItemShared):

When a back forward list item is created set `m_isRootChildFrameItem` and reset
`m_hasPendingBackForwardItem`.

* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/WebProcess/WebPage/WebBackForwardListProxy.cpp:
(WebKit::WebBackForwardListProxy::addItem):
* Source/WebKit/WebProcess/WebPage/WebBackForwardListProxy.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::goToBackForwardItem):

* Source/WebKitLegacy/mac/History/BackForwardList.h:
* Source/WebKitLegacy/mac/History/BackForwardList.mm:
(BackForwardList::addItem):
* Source/WebKitLegacy/mac/History/WebBackForwardList.mm:
(-[WebBackForwardList addItem:]):
* Source/WebKitLegacy/mac/WebView/WebView.mm:
(-[WebView _loadBackForwardListFromOtherView:]):
(-[WebBackForwardList setToMatchDictionaryRepresentation:]):

WebKitLegacy is not supported with site isolation, so we can always just pass the main frame id.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::TEST(SiteIsolation, NavigateIframeBackForward)):

Add a test that navigates an isolated iframe same-origin and verifies that `goBack` and `goForward`
behave correctly.

Canonical link: <a href="https://commits.webkit.org/279152@main">https://commits.webkit.org/279152@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b08933938dc8c5c4caf690c4de4b0b690eee808c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/52428 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/31760 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/4849 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/55702 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/3151 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/54733 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/38255 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/2850 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/42629 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/2019 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/54524 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/29422 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/45254 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/23714 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/26618 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/2529 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/1310 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/48508 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/2679 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/57298 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/27554 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/2716 "Found 1 new test failure: wasm/window-postmessage.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/50018 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/28787 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/45374 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/49270 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/29699 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7725 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/28532 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
```diff
@@ -3124,4 +3124,45 @@ HTTPServer server({
+    HTTPServer server({
+        { "/example"_s, { "<iframe src='https://webkit.org/source'></iframe>"_s } },
+        { "/source"_s, { "<script> alert('done'); </script>"_s } },
+        { "/destination"_s, { "<script> alert('done'); </script>"_s } }
```
May as well make these two alert different strings so it's easier to tell by reading the test what has just finished.

achristensen07 commented on Source/WebKit/UIProcess/WebFrameProxy.h:
```diff
@@ -187,6 +187,9 @@ class WebFrameProxy : public API::ObjectImpl<API::Object::Type::Frame>, public C
     TraversalResult traversePrevious(CanWrap);
 
+    void setHasPendingBackForwardItem() { m_hasPendingBackForwardItem = true; }
+    bool takeHasPendingBackforwardItem() { return std::exchange(m_hasPendingBackForwardItem, false); }
```
nit capital F

achristensen07 reviewed (commented):
Needs more thought.  On my calendar for tomorrow.

achristensen07 commented on Source/WebKit/Shared/WebBackForwardListItem.h:
```diff
@@ -89,6 +89,9 @@ class WebBackForwardListItem : public API::ObjectImpl<API::Object::Type::BackFor
     WebBackForwardCacheEntry* backForwardCacheEntry() const { return m_backForwardCacheEntry.get(); }
     SuspendedPageProxy* suspendedPage() const;
 
+    void setIsRootChildFrameItem(bool rootChildFrameItem) { m_isRootChildFrameItem = rootChildFrameItem; }
```
This seems like a temporary hack to me, but we'll see.  It certainly won't affect site-isolation-off behavior in the meantime.

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -2254,15 +2254,20 @@ RefPtr<API::Navigation> WebPageProxy::goToBackForwardItem(WebBackForwardListItem
 
-    Ref navigation = m_navigationState->createBackForwardNavigation(process().coreProcessIdentifier(), item, m_backForwardList->protectedCurrentItem(), frameLoadType);
+    Ref process = m_process;
+    if (item.url().isNull()) {
```
I don't understand under what conditions this would be hit.  Could we add an ASSERT(preferences().siteIsolationEnabled()) inside that it is for site isolation and catch if it is hit without site isolation on?  Also probably a comment indicating what a null item URL means.

achristensen07 commented on Source/WebKit/UIProcess/WebFrameProxy.h:
```diff
@@ -187,6 +187,9 @@ class WebFrameProxy : public API::ObjectImpl<API::Object::Type::Frame>, public C
     TraversalResult traversePrevious(CanWrap);
 
+    void setHasPendingBackForwardItem() { m_hasPendingBackForwardItem = true; }
+    bool takeHasPendingBackForwardItem() { return std::exchange(m_hasPendingBackForwardItem, false); }
```
I think it would be better to just have setHasPendingBackForwardItem(bool) because you're not really taking anything and the return value of take is not currently used.

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -4661,6 +4668,8 @@ void WebPageProxy::continueNavigationInNewProcess(API::Navigation& navigation, W
             previousProvisionalLoadProcess = Ref { frame.provisionalLoadProcess() }
         ] () mutable {
+            if (!frame->isMainFrame())
+                frame->setHasPendingBackForwardItem();
```
This value is later used to call setIsRootChildFrameItem, but there's no check to see if it is actually a root frame.  It could be a grandchild frame navigating to the domain of its parent, which would hit this code and the frame would not be becoming a root frame.

charliewolfe committed:
[Site Isolation] Begin to fix the back forward list for same-origin iframe navigations
https://bugs.webkit.org/show_bug.cgi?id=274419
rdar://128420789

Reviewed by Alex Christensen.

Some back forward list state for child frames is currently stored only in the web process. This needs to
change with site isolation because web processes can only have history items corresponding to frames they
are hosting or have embedded. With this patch, whenever a root child frame is embedded, we add a separate
item to the back forward list in the UI process. This way, when navigating back or forward, iframe
processes can request history items containing state about same-origin frames that have been embedded by
another process without receiving information about the entire frame tree. We also need to keep track of
which back forward items correspond to initial root child frame loads. This is necessary to determine
which item to navigate an isolated iframe to, while also ignoring these items when the main frame is
moved back or forward.

Further work is required to make the back forward list behave correctly after cross-origin iframe
navigations. More details below.

* Source/WebCore/history/BackForwardClient.h:
* Source/WebCore/history/BackForwardController.cpp:
(WebCore::BackForwardController::addItem):
* Source/WebCore/history/BackForwardController.h:
* Source/WebCore/loader/EmptyClients.cpp:
* Source/WebCore/loader/HistoryController.cpp:
(WebCore::HistoryController::goToItem):
(WebCore::HistoryController::updateBackForwardListClippedAtTarget):
(WebCore::HistoryController::pushState):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::goToItem):
* Source/WebCore/page/Page.h:

Update `addItem` to include a FrameIdentifier parameter so the UI process knows which frame is being
targeted. Also remove several unneeded local frame downcasts.

* Source/WebKit/Shared/WebBackForwardListItem.h:
(WebKit::WebBackForwardListItem::setIsRootChildFrameItem):
(WebKit::WebBackForwardListItem::isRootChildFrameItem const):

Add `m_isRootChildFrameItem` to indicate when a back forward item was created by an initial root child
frame load.

* Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
(WebKit::ProvisionalPageProxy::backForwardAddItem):
* Source/WebKit/UIProcess/ProvisionalPageProxy.h:

* Source/WebKit/UIProcess/WebBackForwardList.cpp:
(WebKit::itemSkippingBackForwardItemsAddedByJSWithoutUserGesture):

Skip root child frame items to avoid targeting initial child frame loads when trying to move the main
frame back/forward. This logic probably still isn’t correct when going back/forward between
cross-origin iframes.

* Source/WebKit/UIProcess/WebFrameProxy.h:
(WebKit::WebFrameProxy::setHasPendingBackForwardItem):
(WebKit::WebFrameProxy::takeHasPendingBackforwardItem):

Add `m_hasPendingBackForwardItem` to indicate when a root frame is waiting for its initial back forward
item to be created.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::goToBackForwardItem):

Change this function to send back forward list items corresponding to isolated iframes to the process
that navigated. `m_lastProcessIdentifier` on `WebBackForwardListItem` is already used for process
selection for PSON. It will initially be the identifier of the web process that constructed the object,
so we can use it for iframe isolation too.

(WebKit::WebPageProxy::didChangeBackForwardList):

Do not call `didChangeBackForwardList` delegates or `setCanGoBack/Forward` for root child frame items to
keep the same behavior as without site isolation.

(WebKit::WebPageProxy::continueNavigationInNewProcess):

Set `m_hasPendingBackForwardItem` to true when creating a new root child frame.

(WebKit::WebPageProxy::backForwardAddItem):
(WebKit::WebPageProxy::backForwardAddItemShared):

When a back forward list item is created set `m_isRootChildFrameItem` and reset
`m_hasPendingBackForwardItem`.

* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/WebProcess/WebPage/WebBackForwardListProxy.cpp:
(WebKit::WebBackForwardListProxy::addItem):
* Source/WebKit/WebProcess/WebPage/WebBackForwardListProxy.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::goToBackForwardItem):

* Source/WebKitLegacy/mac/History/BackForwardList.h:
* Source/WebKitLegacy/mac/History/BackForwardList.mm:
(BackForwardList::addItem):
* Source/WebKitLegacy/mac/History/WebBackForwardList.mm:
(-[WebBackForwardList addItem:]):
* Source/WebKitLegacy/mac/WebView/WebView.mm:
(-[WebView _loadBackForwardListFromOtherView:]):
(-[WebBackForwardList setToMatchDictionaryRepresentation:]):

WebKitLegacy is not supported with site isolation, so we can always just pass the main frame id.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::TEST(SiteIsolation, NavigateIframeBackForward)):

Add a test that navigates an isolated iframe same-origin and verifies that `goBack` and `goForward`
behave correctly.

Canonical link: https://commits.webkit.org/279152@main
